src/pk_core.c

```diff
@@ -648,7 +648,37 @@ DEF(coreExit,
 // String functions.
 // -----------------
 
-// TODO: substring.
+DEF(coreStrSub,
+  "str_sub(str:string, pos:num, len:num) -> string\n"
+  "Returns a substring from a given string supplied. In addition, "
+  "the position and length of the substring are provided when this "
+  "function is called. For example: `str_sub(str, pos, len)`.") {
+  int c = 0;
+  int64_t len, pos;
+
+  String* str = NULL;
+  if (!validateArgString(vm, 1, &str)) return;
+  if (!validateInteger(vm, ARG(2), &pos, "Argument 2")) return;
+  if (!validateInteger(vm, ARG(3), &len, "Argument 3")) return;
+
+  if (str->length < 1)
+    RET_ERR(newString(vm, "Expected a string length more than 0."));
+
+  if (str->length < pos || str->length < len)
+    RET_ERR(newString(vm, "Index out of range."));
+
+  String* subStr = newStringLength(vm, str->data + pos, (uint32_t)len);
+
+  if (subStr == NULL)
+    RET_ERR(newString(vm, "Unable to allocate memory."));
+
+  for (c = 0; c < len; c++) {
+    subStr->data[c] = str->data[pos + c];
+  }
+  subStr->data[c] = '\0';
```

Comment:
The `newStringLength()` will allocate a new string and copy each character using `memcpy()` from the given string. 

And we're passing `str->data + pos` this is where the substring would start, and the `newStringLength` will only copy the given `length` of characters.

Remove this.